### PR TITLE
Use public_suffix for smarter subdomain parsing

### PR DESCRIPTION
**Pull request by humancopy**
Hey :)

This commit makes use of the public_suffix gem to get the subdomain accurately without the need to set the tld_length. This makes it easier running websites that have multiple TLDs.

Peace.

### CODE REVIEW

**Comment by humancopy** [comment]
I've fixed some tests that failed and added support to public_suffix 2. Would love to hear your input about this. I think it's a cleaner solution for parsing domains than having to define the TLD length and other guesswork. ;)

File: lib/apartment/elevators/subdomain.rb [hunk 1]
```diff
@@ -38,13 +43,23 @@ def subdomain(host)
...
-        return [] unless named_host?(host)
+        host_valid?(host) ? parse_host(host) : []
+      end
+
+      def host_valid?(host)
+        not ip_host?(host) and domain_valid?(host)
+      end
+
+      def ip_host?(host)
+        !/^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$/.match(host).nil?
```
Comment (caironoleto) [inline lib/apartment/elevators/subdomain.rb]: This regex match with invalid ips like `333.333.333.33`

File: apartment.gemspec [hunk 2]
```diff
@@ -37,6 +37,7 @@ Gem::Specification.new do |s|
   # must be >= 3.1.2 due to bug in prepared_statements
   s.add_dependency 'activerecord',    '>= 3.1.2', '< 6.0'
   s.add_dependency 'rack',            '>= 1.3.6'
+  s.add_dependency 'public_suffix',   '>= 1.5.3'
```
Comment (caironoleto) [inline apartment.gemspec]: Why this specific version? The latest release of public_suffix is 2.0.5.

File: lib/apartment/elevators/subdomain.rb [hunk 3]
```diff
@@ -6,6 +6,11 @@ module Elevators
     #   Assumes that tenant name should match subdomain
     #
     class Subdomain < Generic
+      def initialize(*args)
+        require 'public_suffix'
```
Comment (caironoleto) [inline lib/apartment/elevators/subdomain.rb]: this require should be after `require 'apartment/elevators/generic'`

File: lib/apartment/elevators/subdomain.rb [hunk 4]
```diff
@@ -38,13 +43,23 @@ def subdomain(host)
...
+        not ip_host?(host) and domain_valid?(host)
+      end
+
+      def ip_host?(host)
+        !/^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$/.match(host).nil?
+      end
 
-        host.split('.')[0..-(Apartment.tld_length + 2)]
+      def domain_valid?(host)
+        PublicSuffix::VERSION[0] == '1' ? PublicSuffix.valid?(host) : PublicSuffix.valid?(host, ignore_private: true)
```
Comment (caironoleto) [inline lib/apartment/elevators/subdomain.rb]: I think if we will add a new dependency, that be the newest version, not the old one.

File: lib/apartment/elevators/subdomain.rb [hunk 5]
```diff
@@ -38,13 +43,23 @@ def subdomain(host)
       end
 
       def subdomains(host)
-        return [] unless named_host?(host)
+        host_valid?(host) ? parse_host(host) : []
+      end
+
+      def host_valid?(host)
+        not ip_host?(host) and domain_valid?(host)
```
Comment (caironoleto) [inline lib/apartment/elevators/subdomain.rb]: avoid `not` and `and`

File: lib/apartment/elevators/subdomain.rb (hunk 1 above)
Comment (humancopy) [inline lib/apartment/elevators/subdomain.rb]: The idea here wasn't to validate an IP but just to check if it's an IP... If it's valid or not I don't think it matters ...

File: apartment.gemspec (hunk 2 above)
Comment (humancopy) [inline apartment.gemspec]: It's just the one I started with last year ... I'll update to latest ;)

Comment (caironoleto) [inline apartment.gemspec]: @humancopy great work! :D

File: lib/apartment/elevators/subdomain.rb (hunk 1 above)
Comment (caironoleto) [inline lib/apartment/elevators/subdomain.rb]: makes sense. At this point, if the host is an IP address it's invalid anyway :)

**Comment by humancopy** [comment]
OK just made a push that uses only public_suffix 2 and cleans the code a bit as per your comments. Let me know if you still find something's off ;)

**Comment by mikecmpbll** [comment]
i think this is good in principle. is there any situation that this strategy would choose an incorrect tld_length for you? what if you wanted to subdomain switch on a private suffix e.g. "client_a.webapp.company.org" ?

i'm just pondering whether there's any merit in leaving the `tld_length` option in as an override but that's my only concern.

**Comment by mikecmpbll** [comment]
ah yeah i just looked in it and i'm happy that it works fine under all those scenarios (sorry, didn't look at the tests properly first time around :)).

File: lib/apartment/elevators/subdomain.rb [hunk 6]
```diff
@@ -7,7 +8,6 @@ module Elevators
     #
     class Subdomain < Generic
       def initialize(*args)
-        require 'public_suffix'
         super
       end
```
Comment (caironoleto) [inline lib/apartment/elevators/subdomain.rb]: Now you can remove this `initialize`

**Pull request merged**